sleap_io/io/nwb.py

```diff
@@ -6,27 +6,298 @@
+    """Converts a SLEAP Instance to an NWB SkeletonInstance.
+
+    Args:
+        instance: A SLEAP Instance.
+
+    Returns:
+        An NWB SkeletonInstance.
+    """
+    if not isinstance(instance, Instance):
+        raise ValueError("The input must be a SLEAP Instance object.")
+    
+    skeleton = slp_skeleton_to_nwb(instance.skeleton)
+    points_list = list(instance.points.values())
+    node_locs = [[point.x, point.y] for point in points_list]
+    np_node_locations = np.array(node_locs)
+    return SkeletonInstance(
+        name=name_generator("skeleton_instance"),
+        id=np.uint(10),
+        node_locations=np_node_locations,
+        node_visibility=[point.visible for point in instance.points.values()],
+        skeleton=skeleton,
+    )
+
+
+def videos_to_source_videos(videos: List[Video]) -> SourceVideos:  # type: ignore[return]
+    """Converts a list of SLEAP Videos to NWB SourceVideos.
+
+    Args:
+        videos: A list of SLEAP Videos.
+
+    Returns:
+        An NWB SourceVideos object.
+    """
+    if not isinstance(videos, list) or not all(isinstance(video, Video) for video in videos):
+        raise ValueError("The input must be a list of SLEAP Video objects.")
+    
+    source_videos = []
+    for video in videos:
+        image_series = ImageSeries(
+            name=name_generator("video"),
+            description="Video file",
+            unit="NA",
+            format="external",
+            external_file=[video.filename],
+            dimension=[video.backend.img_shape[0], video.backend.img_shape[1]],
+            starting_frame=[0],
+            rate=30.0,  # TODO - change to `video.backend.fps` when available
+        )
+        source_videos.append(image_series)
+    return SourceVideos(image_series=source_videos)
+
+
+def sleap_pkg_to_nwb(filename: str, **kwargs) -> NWBFile:
+    """Write a SLEAP package to an NWB file.
+
+    Args:
+        filename: The path to the SLEAP package.
+
+    Returns:
+        An NWBFile object.
+    """
+    if not filename.endswith(".pkg.slp"):
+        raise ValueError("The filename must end with '.pkg.slp'.")
+
+    labels = load_slp(filename)
+
+    save_path = Path(filename.replace(".slp", ".nwb"))
+    save_path.mkdir(parents=True, exist_ok=True)
+    img_paths = []
+    for i, labeled_frame in enumerate(labels.labeled_frames):
+        img_path = save_path / f"frame_{i}.png"
+        if labeled_frame.image.ndim == 3 and labeled_frame.image.shape[-1] == 1:
+            imwrite(img_path, labeled_frame.image[:, :, 0])
+        else:
+            imwrite(img_path, labeled_frame.image)
+        img_paths.append(img_path)
+    
+    # then use img_paths when saving the NWB TrainingFrames with references
+    # to the appropriate image files
+    save_nwb(labels, save_path, img_paths=img_paths, **kwargs)
+    raise NotImplementedError("This function is not yet implemented.")
```

Comment:
This whole function is not how we want to do this.

`save_nwb` is the high level function and should be the one calling your lower level functions. Do not import things from `sleap_io.io.main` in this module at all!

When calling `save_nwb(..., as_training=True)`, that function should call the appropriate functions within this module.

That call will have the `labels` and the `save_path`, allowing you to pass those on to the image writer method.

When image writing (should be its own function), follow the same logic used in `sleap_io.io.slp` to select the appropriate image writing backend, i.e.:

1. Try to import OpenCV:

    https://github.com/talmolab/sleap-io/blob/5b79a2296572b6e10a75606020d997320e16ecf3/sleap_io/io/slp.py#L30-L36

2. Fallback to the appropriate API by checking for the module:

    https://github.com/talmolab/sleap-io/blob/5b79a2296572b6e10a75606020d997320e16ecf3/sleap_io/io/slp.py#L229-L239

(Here we don't need the in-memory buffer logic since we're saving out to files.)

We should also save to a folder with the same name as the NWB file, but without the extension. It'll create conflicts if the file and folder have the exact same name, so just drop the suffix.

After saving the images, construct an `ImageSeries` and keep track of the indexing metadata (`(video, frame_idx) -> img_path`) so you can pass it around to the other functions without creating duplicates.

Image saving should be optional in case we want to save an NWB file with simple pointers to the source video files, but either way we should store the metadata about the source videos in one of the NWB attributes so we can recreate the source video objects on load.